### PR TITLE
feat(new sink): Initial `databend` sink

### PR DESCRIPTION
**Pull request by everpcpc**
fix: #15727

need: https://github.com/datafuselabs/databend/issues/9569

### CODE REVIEW

**Comment by spencergilbert** [comment]
Just started reviewing this, and playing with it locally @everpcpc

The integration test is currently failing:
```
failures:
    sinks::databend::integration_tests::insert_events

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 245 filtered out; finished in 0.18s


--- TRY 1 STDERR:        vector sinks::databend::integration_tests::insert_events ---
thread 'sinks::databend::integration_tests::insert_events' panicked at 'Failed to assert compliance, errors:
  - Missing event `BytesSent`
  - Missing metric `component_sent_bytes_total{endpoint,protocol}`
', src/test_util/components.rs:427:16
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

As we're missing some required telemetry (ref: [ComponentBytesSent](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#componentbytessent))

**Comment by everpcpc** [comment]
It seems `Integration - Linux, datadog-*` are always failing since they cannot get secrets when running on forks.
Maybe we should separate these to another workflow?

**Comment by spencergilbert** [comment]
> It seems `Integration - Linux, datadog-*` are always failing since they cannot get secrets when running on forks. Maybe we should separate these to another workflow, or ignore them on forks?

Yeah, ☹️ it's on our todo list to run them conditionally but we haven't gotten to it.

**Review by spencergilbert — COMMENTED** [review]
Mhmm, developing locally on Darwin - the integration tests are failing for me still.

```text
thread 'sinks::databend::integration_tests::insert_events' panicked at 'called `Result::unwrap()` on an `Err` value: Client { message: "Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 111)" }', src/sinks/databend/integration_tests.rs:87:10
```

This one seems like we're trying to connect to Databend before it's ready, we probably want to probe the endpoint until it's up before starting the tests.

```text
thread 'sinks::databend::integration_tests::insert_events' panicked at 'called `Result::unwrap()` on an `Err` value: Server { code: 401, message: "Http Status not OK" }', src/sinks/databend/integration_tests.rs:87:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This one seems like the user/pass isn't getting used/setup properly?

File: src/sinks/databend/service.rs
Comment (spencergilbert) [inline src/sinks/databend/service.rs:1]: Can we move this into the second group of third party crates?

File: src/sinks/databend/service.rs
Comment (spencergilbert) [inline src/sinks/databend/service.rs:109]:
```suggestion
        let now = Utc::now().timestamp();
```

I think we generally use `Utc` - is there a particular reason for `SystemTime` here?

File: src/sinks/databend/service.rs
Comment (spencergilbert) [inline src/sinks/databend/service.rs:119]:
```suggestion
            now,
```

File: src/sinks/databend/service.rs
Comment (spencergilbert) [inline src/sinks/databend/service.rs:169]: These options are always hardcoded, should they be static instead of cloning?

File: src/sinks/databend/sink.rs
Comment (spencergilbert) [inline src/sinks/databend/sink.rs:61]: Can you clarify how this isn't being used?

File: src/sinks/databend/api.rs
Comment (spencergilbert) [inline src/sinks/databend/api.rs:62]: Does this need to be signed, could it just be `usize` or `u32`?

File: src/sinks/databend/api.rs
Comment (spencergilbert) [inline src/sinks/databend/api.rs:76]: If we're not going to use these commented out fields can we remove them until they're needed?

It doesn't look like a few of these are used anywhere, should we wait until there's a need to include them?

File: src/sinks/databend/api.rs
Comment (spencergilbert) [inline src/sinks/databend/api.rs:67]: Can we mark these as `pub (super)` if they're only used in this sink.

**Review by spencergilbert — CHANGES_REQUESTED** [review]
Hey! Couple more notes after a second review.

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs]: fixed

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs]: fixed

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs]: changed to `u16`

File: src/sinks/databend/sink.rs
Comment (everpcpc) [reply src/sinks/databend/sink.rs:73]: fixed in `events_sent` of `DriverResponse`

File: src/sinks/databend/service.rs
Comment (everpcpc) [reply src/sinks/databend/service.rs]: made them as lazy load static

File: src/sinks/databend/service.rs
Comment (everpcpc) [reply src/sinks/databend/service.rs]: fixed

File: src/sinks/databend/service.rs
Comment (everpcpc) [reply src/sinks/databend/service.rs]: no particular reason, changed to `Utc`

File: src/sinks/databend/service.rs
Comment (everpcpc) [reply src/sinks/databend/service.rs]: fixed

File: src/sinks/databend/config.rs
Comment (everpcpc) [reply src/sinks/databend/config.rs:135]: Maybe we can consider adding it down the line.

File: src/sinks/databend/config.rs
Comment (everpcpc) [reply src/sinks/databend/config.rs]: fixed

File: src/sinks/databend/config.rs
Comment (everpcpc) [reply src/sinks/databend/config.rs]: Yes, we only need `endpoint`, alias removed.

**Comment by everpcpc** [comment]
> we probably want to probe the endpoint until it's up before starting the tests.

Sure, I added a `healthcheck` for the `databend` service in docker compose file.

File: src/sinks/databend/event_encoder.rs
Comment (everpcpc) [reply src/sinks/databend/event_encoder.rs:10]: Yes! the existing `json` codec can be used. But I'm also considering adding a `csv` format support.
Is it OK to add a `csv` to the existing `lib/codecs`?

Comment (spencergilbert) [reply src/sinks/databend/event_encoder.rs:10]: Yeah, that's been an open feature request https://github.com/vectordotdev/vector/issues/7124 - we'd happy take a PR adding a new encoder for CSV.

Would there be a user-visible difference between encoding the events as `json` vs `csv`, I feel like if it's not impacting the user it may not be worth exposing as a configuration option?

File: src/sinks/databend/api.rs
Comment (spencergilbert) [inline src/sinks/databend/api.rs:84]: Are these fields that are `allow(dead_code)` actually needed, it didn't look like they were getting used anywhere?

**Review by spencergilbert — COMMENTED** [review]
I'm still having trouble with the integration tests locally, not sure where the issue lies but I pulled these logs from Databend:

```jsonl
{"timestamp":"2023-02-01T17:04:50.867646Z","level":"INFO","fields":{"message":"run query_id=57b8d53a-0a35-4936-b9e0-6f00f468b67a in session_id=ce22ca99-cd63-40ed-8d84-23571206f3d8, sql='PRESIGN UPLOAD @~/vector/default/test_dc9mabu7rq/1675271090-QBDcWUig'"},"target":"databend_query::servers::http::v1::query::http_query"}
{"timestamp":"2023-02-01T17:04:50.867940Z","level":"INFO","fields":{"message":"http query 57b8d53a-0a35-4936-b9e0-6f00f468b67a, change state to Running"},"target":"databend_query::servers::http::v1::query::execute_state"}
{"timestamp":"2023-02-01T17:04:50.867984Z","level":"INFO","fields":{"message":"{\"log_type\":1,\"handler_type\":\"HTTPQuery\",\"tenant_id\":\"default\",\"cluster_id\":\"default\",\"sql_user\":\"vector\",\"query_id\":\"57b8d53a-0a35-4936-b9e0-6f00f468b67a\",\"query_kind\":\"Presign\",\"query_text\":\"PRESIGN UPLOAD @~/vector/default/test_dc9mabu7rq/1675271090-QBDcWUig\",\"event_date\":\"2023-02-01\",\"event_time\":\"2023-02-01 17:04:50.867958\",\"query_start_time\":\"2023-02-01 17:04:50.867643\",\"query_duration_ms\":0,\"current_database\":\"default\",\"databases\":\"\",\"tables\":\"\",\"columns\":\"\",\"projections\":\"\",\"written_rows\":0,\"written_bytes\":0,\"written_io_bytes\":0,\"written_io_bytes_cost_ms\":0,\"scan_rows\":0,\"scan_bytes\":0,\"scan_io_bytes\":0,\"scan_io_bytes_cost_ms\":0,\"scan_partitions\":0,\"total_partitions\":0,\"result_rows\":0,\"result_bytes\":0,\"cpu_usage\":6,\"memory_usage\":0,\"client_info\":\"\",\"client_address\":\"\",\"exception_code\":0,\"exception_text\":\"\",\"stack_trace\":\"\",\"server_version\":\"\",\"extra\":\"\"}"},"target":"databend_query::interpreters::interpreter_query_log"}
{"timestamp":"2023-02-01T17:04:50.868291Z","level":"WARN","fields":{"message":"receiver has been disconnected, finish executor now"},"target":"databend_query::pipelines::executor::pipeline_pulling_executor"}
{"timestamp":"2023-02-01T17:04:50.868323Z","level":"INFO","fields":{"message":"http query 57b8d53a-0a35-4936-b9e0-6f00f468b67a got new block with 1 rows"},"target":"databend_query::servers::http::v1::query::page_manager"}
{"timestamp":"2023-02-01T17:04:50.868349Z","level":"INFO","fields":{"message":"{\"log_type\":2,\"handler_type\":\"HTTPQuery\",\"tenant_id\":\"default\",\"cluster_id\":\"default\",\"sql_user\":\"vector\",\"query_id\":\"57b8d53a-0a35-4936-b9e0-6f00f468b67a\",\"query_kind\":\"Presign\",\"query_text\":\"PRESIGN UPLOAD @~/vector/default/test_dc9mabu7rq/1675271090-QBDcWUig\",\"event_date\":\"2023-02-01\",\"event_time\":\"2023-02-01 17:04:50.868313\",\"query_start_time\":\"2023-02-01 17:04:50.867643\",\"query_duration_ms\":0,\"current_database\":\"default\",\"databases\":\"\",\"tables\":\"\",\"columns\":\"\",\"projections\":\"\",\"written_rows\":0,\"written_bytes\":0,\"written_io_bytes\":0,\"written_io_bytes_cost_ms\":0,\"scan_rows\":0,\"scan_bytes\":0,\"scan_io_bytes\":0,\"scan_io_bytes_cost_ms\":0,\"scan_partitions\":0,\"total_partitions\":0,\"result_rows\":1,\"result_bytes\":24,\"cpu_usage\":6,\"memory_usage\":0,\"client_info\":\"\",\"client_address\":\"\",\"exception_code\":0,\"exception_text\":\"\",\"stack_trace\":\"\",\"server_version\":\"\",\"extra\":\"\"}"},"target":"databend_query::interpreters::interpreter_query_log"}
{"timestamp":"2023-02-01T17:04:50.868406Z","level":"INFO","fields":{"message":"http query 57b8d53a-0a35-4936-b9e0-6f00f468b67a: change state to Stopped, reason Ok(())"},"target":"databend_query::servers::http::v1::query::execute_state"}
{"timestamp":"2023-02-01T17:04:50.868437Z","level":"INFO","fields":{"message":"http query 57b8d53a-0a35-4936-b9e0-6f00f468b67a reach end of blocks"},"target":"databend_query::servers::http::v1::query::page_manager"}
{"timestamp":"2023-02-01T17:04:50.868452Z","level":"INFO","fields":{"message":"initial response to http query_id=57b8d53a-0a35-4936-b9e0-6f00f468b67a, state=ResponseState { running_time_ms: 0.7622920000000001, progresses: Progresses { scan_progress: ProgressValues { rows: 0, bytes: 0 }, write_progress: ProgressValues { rows: 0, bytes: 0 }, result_progress: ProgressValues { rows: 1, bytes: 24 } }, state: Succeeded, affect: None, error: None }, rows=1, next_page=None, sql='PRESIGN UPLOAD @~/vector/default/test_dc9mabu7rq/1675271090-QBDcWUig'"},"target":"databend_query::servers::http::v1::http_query_handlers"}
{"timestamp":"2023-02-01T17:04:50.869154Z","level":"INFO","fields":{"message":"receive http handler request: Request { method: POST, uri: /v1/query/57b8d53a-0a35-4936-b9e0-6f00f468b67a/final, version: HTTP/1.1, headers: {\"content-type\": \"application/json\", \"authorization\": \"Basic dmVjdG9yOnZlY3Rvcg==\", \"user-agent\": \"Vector/0.28.0 (aarch64-unknown-linux-gnu debug=full)\", \"accept-encoding\": \"identity\", \"host\": \"databend:8000\"} },"},"target":"databend_query::servers::http::middleware"}
```

And these logs from Vector:

```text
2023-02-01T17:04:50.872333Z ERROR vector::sinks::util::retries: Non-retriable error; dropping the request. error=Server responded with an error: 507, Presigned Upload Failed internal_log_rate_limit=true
2023-02-01T17:04:50.872458Z ERROR vector_common::internal_event::service: Service call failed. No retries or retries exhausted. error=Some(Server { code: 507, message: "Presigned Upload Failed" }) request_id=1 error_type="request_failed" stage="sending" internal_log_rate_limit=true
2023-02-01T17:04:50.872566Z ERROR vector_common::internal_event::component_events_dropped: Events dropped intentional=false count=1 reason="Service call failed. No retries or retries exhausted." internal_log_rate_limit=true
```

The test also returns that it's missing the `BytesSent` event, but that may be due to the lack of successful sending.

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs:84]: They are used in integration test currently.

**Comment by everpcpc** [comment]
> Server { code: 507, message: "Presigned Upload Failed" }

This error seems to be from MinIO. I could add a more detailed log to see what happened.

**Comment by spencergilbert** [comment]
> > Server { code: 507, message: "Presigned Upload Failed" }
> 
> This error seems from MinIO. I could add a more detailed log to see what happened. Maybe disk full I guess ?

Ah, for the life of me I can't find how to get logs out of that minio container 😅 but using a volume for that `/data` dir solves the problem for me. I'm not sure why that happens for me locally vs yours locally vs CI 😆 but assuming Docker on Mac problems.

File: src/sinks/databend/sink.rs
Comment (spencergilbert) [reply src/sinks/databend/sink.rs:73]:
```suggestion
                    event_collection.events_byte_size, // this event is already json encoded
```

This would be something to keep in mind if a `csv` encoder is also desired down the line, since it wouldn't necessarily be the case.

**Comment by everpcpc** [comment]
> but assuming Docker on Mac problems.

I use [colima](https://github.com/abiosoft/colima) on macOS, and also encountered similar problem.
But my situation is that I set the colima disk size limit to 60GiB and it was full. After cleaning up some old docker images it recoverd. 😄 

```
❯ colima list
PROFILE    STATUS     ARCH       CPUS    MEMORY    DISK     RUNTIME    ADDRESS
default    Stopped    aarch64    4       8GiB      60GiB
```

**Comment by spencergilbert** [comment]
😅 I wish docker for mac was easier to use... 
image


Might need to take another look at colima.

File: src/sinks/databend/api.rs
Comment (spencergilbert) [reply src/sinks/databend/api.rs:84]: We'd like to avoid the `dead_code` if possible, whether that's putting it behind a `cfg(test)` or just using those outside of the tests as well.

**Review by spencergilbert — CHANGES_REQUESTED** [review]
Thanks again, this is definitely coming along and almost there. The main thing being using the `json` codec for event encoding.

File: website/cue/reference/components/sinks/base/databend.cue
Comment (jszwedko) [inline website/cue/reference/components/sinks/base/databend.cue:189]: Seems like we have some extra horizontal whitespace in this file 😅

Comment (spencergilbert) [reply website/cue/reference/components/sinks/base/databend.cue:189]: I believe that's the case for _all_ generated cue files, they're valid just wonky

Comment (jszwedko) [reply website/cue/reference/components/sinks/base/databend.cue:189]: Actually, I see this is coming from the generation so ignore me. Seems like something we need to fix at a deeper level.

Comment (jszwedko) [reply website/cue/reference/components/sinks/base/databend.cue:189]: > I believe that's the case for _all_ generated cue files, they're valid just wonky

Indeed, just realized that :)

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs]: changed to `#[cfg(all(test, feature = "databend-integration-tests"))]`

**Comment by everpcpc** [comment]
Hi, I've refactored the encoding part with `RequestBuilder`.

**Comment by spencergilbert** [comment]
Just noting I plan on reviewing this again tomorrow, still on my radar!

File: src/sinks/databend/config.rs
Comment (spencergilbert) [inline src/sinks/databend/config.rs:122]:
```suggestion
        let encoding_config = EncodingConfigWithFraming::new(
            Some(FramingConfig::NewlineDelimited),
            SerializerConfig::Json(JsonSerializerConfig),
            self.encoding,
        );
        let (framer, serializer) = encoding_config.build(SinkType::StreamBased)?;
```

Continuing off the above suggestion, you should be able to do something like this to build the encoding configuration from the user provided `Transformer` and internally selecting Json and newline delimiters (`None` for the framer should [do this regardless](https://github.com/vectordotdev/vector/blob/master/src/codecs/encoding/config.rs#L103-L104), for `StreamBased` but why not by specific).

File: src/sinks/databend/config.rs
Comment (spencergilbert) [inline src/sinks/databend/config.rs:124]: The suggested changes would also let you set these without matching.

File: src/sinks/databend/config.rs
Comment (everpcpc) [reply src/sinks/databend/config.rs]: removed

File: src/sinks/databend/config.rs
Comment (everpcpc) [reply src/sinks/databend/config.rs]: fixed

File: src/sinks/databend/config.rs
Comment (everpcpc) [reply src/sinks/databend/config.rs]: removed

**Review by spencergilbert — APPROVED** [review]
Sorry, I thought I had approved this earlier. @davidhuie-dd if you wanted take a pass we can get this merged!

File: src/sinks/databend/api.rs
Comment (davidhuie-dd) [inline src/sinks/databend/api.rs:111]: Re terminology: wouldn't this method be referring to the endpoint and the `get_endpoint` method above be referring to the host?

File: src/sinks/databend/integration_tests.rs
Comment (davidhuie-dd) [inline src/sinks/databend/integration_tests.rs:113]: nit: s/actrual/actual/

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs:138]: Sure.

File: src/sinks/databend/api.rs
Comment (everpcpc) [reply src/sinks/databend/api.rs:199]: OK

File: src/sinks/databend/integration_tests.rs
Comment (davidhuie-dd) [reply src/sinks/databend/integration_tests.rs]: Looks like this one is still there. Happy to approve once this one is fixed 😄

Comment (everpcpc) [reply src/sinks/databend/integration_tests.rs]: Oh sorry I missed that.

**Comment by spencergilbert** [comment]
Thanks so much for working through this and going back and forth on this with us @everpcpc ❤️

**Comment by everpcpc** [comment]
> Thanks so much for working through this and going back and forth on this with us @everpcpc ❤️

Also thanks so much for the continuously attention with carefully code review and helping on this ❤️

**Pull request merged**
